x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java

```diff
@@ -793,41 +797,43 @@ private void invokeRollover(
         final Executor executor = threadPool.generic();
         executor.execute(() -> {
             List<String> rolloverTargets = findRolloverTargetDataStreams(state, templateName, indexTemplate);
-            if (rolloverTargets.isEmpty() == false) {
-                GroupedActionListener<RolloverResponse> groupedActionListener = new GroupedActionListener<>(
-                    rolloverTargets.size(),
-                    new ActionListener<>() {
-                        @Override
-                        public void onResponse(Collection<RolloverResponse> rolloverResponses) {
-                            creationCheck.set(false);
-                            onRolloversBulkResponse(rolloverResponses);
-                        }
+            if (rolloverTargets.isEmpty()) {
+                creationCheck.set(false);
+                return;
+            }
```

Comment:
This fixes another bug: we were never setting creationCheck to false in the case where there were no data streams to roll over on upgrade. This led to the plugin not recreating templates as intended, as the registry thought template creation was perpetually in-progress.

Reply:
BTW without this change, the existing YAML REST tests would fail with the change to trigger the rollover path on template creation. I've also added a test case for template reinstallation for good measure.

Reply:
Nice catch! I would recommend to change the code a bit to make this less error prone. I think it will be easier if the responsibility to reset the check flag would happen in one spot. I think it would be nice to move to https://github.com/elastic/elasticsearch/pull/116219/files#diff-a74931128493bfa2cd8c35c1bf3378014fd64c596c4bfbea0f91ddedd5db7011R552-R555.

For example:

```
 if (applyRolloverAfterTemplateV2Update()) {
       invokeRollover(state, templateName, indexTemplate, () -> creationCheck.set(false));
} else {
     creationCheck.set(false);
}
```

Then the `invokeRollover` would apply it accordingly:
```
    private void invokeRollover(
        final ClusterState state,
        final String templateName,
        final ComposableIndexTemplate indexTemplate,
        Runnable runAfter
    ) {
        final Executor executor = threadPool.generic();
        executor.execute(() -> {
            List<String> rolloverTargets = findRolloverTargetDataStreams(state, templateName, indexTemplate);
            if (rolloverTargets.isEmpty()) {
                runAfter.run();
                return;
            }
            ActionListener<RolloverResponse> groupedActionListener = ActionListener.runAfter(new GroupedActionListener<>(
                rolloverTargets.size(),
                new ActionListener<>() {
                    @Override
                    public void onResponse(Collection<RolloverResponse> rolloverResponses) {
                        onRolloversBulkResponse(rolloverResponses);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        onRolloverFailure(e);
                    }
                }), runAfter);
            for (String rolloverTarget : rolloverTargets) {
                logger.info(
                    "rolling over data stream [{}] lazily as a followup to the upgrade of the [{}] index template [{}]",
                    rolloverTarget,
                    getOrigin(),
                    templateName
                );
                RolloverRequest request = new RolloverRequest(rolloverTarget, null);
                request.lazy(true);
                request.masterNodeTimeout(TimeValue.MAX_VALUE);
                executeAsyncWithOrigin(
                    client.threadPool().getThreadContext(),
                    getOrigin(),
                    request,
                    groupedActionListener,
                    (req, listener) -> client.execute(RolloverAction.INSTANCE, req, listener)
                );
            }
        });
    }
```

The benefit of the above, is that `creationCheck` would be reset once when all the listeners have finished (assuming my understanding is correct) and the responsibility lies on the first snippet to declare that there is something that needs to be run and the `invokeRollover` needs to apply that `runAfter`, whatever that is.

Does this make sense? By the way this is a draft, if you like the suggestion please test it also ;)

Reply:
:+1: I like it, thanks for the suggestion. I agree that it's a bit error prone (evidently!) with the way it's spread around at the moment. I will test and push an update shortly.